content/en/blog/_posts/2023-03-08-memory-qos-cgroups-v2/index.md

```diff
@@ -0,0 +1,260 @@
+        <br>
+        <br>
+        <i>We map it to the container's memory limit</i>
+        </td>
+   </tr>
+   <tr>
+        <td>memory.min</td>
+        <td><code>memory.min</code> specifies a minimum amount of memory the cgroup must always retain, i.e., memory that can never be reclaimed by the system.
+        If there's no unprotected reclaimable memory available, OOM kill is invoked.
+        <br>
+        <br>
+        <i>We map it to the container's memory request</i>
+        </td>
+   </tr>
+   <tr>
+       <td>memory.high</td>
+       <td><code>memory.high</code> is the memory usage throttle limit. This is the main mechanism to control a cgroup's memory use. If a cgroup's memory use goes over the high boundary specified here, the cgroup’s processes are throttled and put under heavy reclaim pressure. The default is maximum, meaning there is no limit. 
+       <br>
+       <br>
+       <i>We use a formula to calculate <code>memory.high</code>, depending on container's memory request, memory limit or node allocatable memory (if container's memory limit is empty) and a throttling factor. Please refer to the KEP for more details on the formula.</i>
+       </td>
+   </tr>
+</table>
+
+**Note**: If a container has no memory limit specified, we substitute limits.memory for node allocatable memory.
+
+### Support for Pod QoS classes
+Based on user feedback for the Alpha feature in Kubernetes v1.22, some users would like 
+to opt out of MemoryQoS on a per-pod basis to ensure there is no early memory throttling.
+Therefore, in Kubernetes v1.27 Memory QOS also supports memory.high to be set as per 
+Quality of Service(QoS) for Pod classes. Following are the different cases for memory.high
+as per QOS classes:
+
+1. **Guaranteed pods** by their QoS definition require memory requests=memory limits and are 
+not overcommitted. Hence MemoryQoS feature is disabled on those pods by not setting 
+memory.high. This ensures that Guaranteed pods can fully use their memory requests up
+to their set limit, and not hit any throttling.
+
+2. **Burstable pods** by their QoS definition require at least one container in the Pod with
+CPU or memory request or limit set.
+    
+    * When requests.memory and limits.memory are set, the formula is used as-is:
+    ![](./container-memory-high-limit.svg)  
+    <sub>i: the i<sup>th</sup> container in one pod</sub>
+
+    * When requests.memory is set, limits.memory is not set, we substitute limits.memory
+    for node allocatable memory in the formula:
+    ![](./container-memory-high-no-limits.svg)  
+    <sub>i: the i<sup>th</sup> container in one pod</sub>
+
+3. BestEffort The pod gets a BestEffort class if limits.memory and requests.memory are not 
+set. We set requests.memory = 0 and substitute limits.memory for node allocatable memory 
+in the formula:
+
+    ![](./container-memory-high-best-effort.svg)  
+    <sub>i: the i<sup>th</sup> container in one pod</sub>
+
```

Comment:
It might be worth to summarize here - the summary is only pods in burstable and best effort QoS will set `memory.high`. Guaranteed QoS pods do not set `memory.high` because their memory is guaranteed.